File: src/packages/control-ui/src/api/quic_connection_controller.ts
export const QuicConnectionState = {
  Idle: 'Idle',
  Initial: 'Initial',
  Handshaking: 'Handshaking',
  Established: 'Established',
  Closing: 'Closing',
  Closed: 'Closed',
} as const;
export type QuicConnectionState = (typeof QuicConnectionState)[keyof typeof QuicConnectionState];

export interface QuicConnectionMetrics {
  state: QuicConnectionState;
  smoothedRttMs: number;
  rttvarMs: number;
  minRttMs: number;
  congestionWindowBytes: number;
  bytesInFlight: number;
  lostPacketsCount: number;
}

interface SentPacketRecord {
  packetNumber: number;
  sizeBytes: number;
  timeSentMs: number;
}

export class QuicConnectionController {
  private state: QuicConnectionState = QuicConnectionState.Idle;
  private smoothedRttMs: number = 100;
  private rttvarMs: number = 50;
  private minRttMs: number = Number.MAX_SAFE_INTEGER;
  private cwndBytes: number;
  private bytesInFlight: number = 0;
  private lostPacketsCount: number = 0;
  private ssthreshBytes: number = Number.MAX_SAFE_INTEGER;
  private sentPackets: Map<number, SentPacketRecord> = new Map();

  constructor(initialWindowBytes: number = 20000) {
    this.cwndBytes = Math.max(14720, initialWindowBytes);
  }

  setState(state: QuicConnectionState): void {
    this.state = state;
  }

  canSend(packetBytes: number): boolean {
    return (
      this.state === QuicConnectionState.Established &&
      this.bytesInFlight + packetBytes <= this.cwndBytes
    );
  }

  onPacketSent(packetNumber: number, sizeBytes: number, timeSentMs: number): void {
    this.bytesInFlight += sizeBytes;
    this.sentPackets.set(packetNumber, {
      packetNumber,
      sizeBytes,
      timeSentMs,
    });
  }

  onAckReceived(ackedPacketNumber: number, nowMs: number): void {
    const record = this.sentPackets.get(ackedPacketNumber);
    if (!record) return;

    this.sentPackets.delete(ackedPacketNumber);
    this.bytesInFlight = Math.max(0, this.bytesInFlight - record.sizeBytes);

    if (nowMs >= record.timeSentMs) {
      const sampleRtt = nowMs - record.timeSentMs;
      this.updateRtt(sampleRtt);
    }

    // Congestion window expansion
    if (this.cwndBytes < this.ssthreshBytes) {
      this.cwndBytes += record.sizeBytes;
    } else {
      const increment = Math.floor((1472 * 1472) / Math.max(1, this.cwndBytes));
      this.cwndBytes += Math.max(1, increment);
    }
  }

  onPacketLoss(lostPacketNumber: number): void {
    const record = this.sentPackets.get(lostPacketNumber);
    if (!record) return;

    this.sentPackets.delete(lostPacketNumber);
    this.bytesInFlight = Math.max(0, this.bytesInFlight - record.sizeBytes);
    this.lostPacketsCount += 1;

    // Multiplicative decrease
    this.ssthreshBytes = Math.max(14720, Math.floor(this.cwndBytes / 2));
    this.cwndBytes = this.ssthreshBytes;
  }

  calculatePtoMs(): number {
    return this.smoothedRttMs + Math.max(10, 4 * this.rttvarMs);
  }

  getMetrics(): QuicConnectionMetrics {
    return {
      state: this.state,
      smoothedRttMs: this.smoothedRttMs,
      rttvarMs: this.rttvarMs,
      minRttMs: this.minRttMs === Number.MAX_SAFE_INTEGER ? 0 : this.minRttMs,
      congestionWindowBytes: this.cwndBytes,
      bytesInFlight: this.bytesInFlight,
      lostPacketsCount: this.lostPacketsCount,
    };
  }

  private updateRtt(sampleRtt: number): void {
    if (this.minRttMs === Number.MAX_SAFE_INTEGER) {
      this.minRttMs = sampleRtt;
      this.smoothedRttMs = sampleRtt;
      this.rttvarMs = Math.floor(sampleRtt / 2);
    } else {
      this.minRttMs = Math.min(this.minRttMs, sampleRtt);
      const diff = Math.abs(sampleRtt - this.smoothedRttMs);
      this.rttvarMs = Math.floor((3 * this.rttvarMs + diff) / 4);
      this.smoothedRttMs = Math.floor((7 * this.smoothedRttMs + sampleRtt) / 8);
    }
  }
}

File: labs/mobile/donor-archive/tunnel/QuicConnectionController.kt
package com.luminet.android.tunnel

import kotlin.math.abs
import kotlin.math.max

enum class QuicConnectionState {
    IDLE,
    HANDSHAKING,
    ESTABLISHED,
    CLOSING,
    CLOSED
}

data class QuicConnectionMetrics(
    val state: QuicConnectionState,
    val smoothedRttMs: Int,
    val rttVarMs: Int,
    val minRttMs: Int,
    val cwndBytes: Long,
    val bytesInFlight: Long,
    val lostPacketCount: Long
)

class QuicConnectionController(initialWindowBytes: Long = 14720) {
    var state: QuicConnectionState = QuicConnectionState.IDLE
    var smoothedRttMs: Int = 100
        private set
    var rttVarMs: Int = 50
        private set
    var minRttMs: Int = Int.MAX_VALUE
        private set
    var cwndBytes: Long = initialWindowBytes.coerceAtLeast(14720)
        private set
    var bytesInFlight: Long = 0
        private set
    var lostPacketCount: Long = 0
        private set
    private var ssthreshBytes: Long = Long.MAX_VALUE

    private data class SentRecord(val size: Int, val timeSentMs: Long)
    private val sentPackets = HashMap<Long, SentRecord>()

    fun canSend(size: Int): Boolean {
        return state == QuicConnectionState.ESTABLISHED && (bytesInFlight + size <= cwndBytes)
    }

    fun onPacketSent(pktNum: Long, size: Int, timeSentMs: Long) {
        bytesInFlight += size
        sentPackets[pktNum] = SentRecord(size, timeSentMs)
    }

    fun onAckReceived(pktNum: Long, nowMs: Long) {
        val rec = sentPackets.remove(pktNum) ?: return
        bytesInFlight = (bytesInFlight - rec.size).coerceAtLeast(0)

        if (nowMs >= rec.timeSentMs) {
            val sampleRtt = (nowMs - rec.timeSentMs).toInt()
            updateRtt(sampleRtt)
        }

        if (cwndBytes < ssthreshBytes) {
            cwndBytes += rec.size
        } else {
            val inc = (1472L * 1472L) / cwndBytes
            cwndBytes += max(1L, inc)
        }
    }

    fun onPacketLoss(pktNum: Long) {
        val rec = sentPackets.remove(pktNum) ?: return
        bytesInFlight = (bytesInFlight - rec.size).coerceAtLeast(0)
        lostPacketCount++

        ssthreshBytes = max(14720L, cwndBytes / 2)
        cwndBytes = ssthreshBytes
    }

    fun calculatePtoMs(): Int {
        return smoothedRttMs + max(10, 4 * rttVarMs)
    }

    fun getMetrics(): QuicConnectionMetrics {
        return QuicConnectionMetrics(
            state = state,
            smoothedRttMs = smoothedRttMs,
            rttVarMs = rttVarMs,
            minRttMs = if (minRttMs == Int.MAX_VALUE) 0 else minRttMs,
            cwndBytes = cwndBytes,
            bytesInFlight = bytesInFlight,
            lostPacketCount = lostPacketCount
        )
    }

    private fun updateRtt(sample: Int) {
        if (minRttMs == Int.MAX_VALUE) {
            minRttMs = sample
            smoothedRttMs = sample
            rttVarMs = sample / 2
        } else {
            if (sample < minRttMs) minRttMs = sample
            val diff = abs(sample - smoothedRttMs)
            rttVarMs = (3 * rttVarMs + diff) / 4
            smoothedRttMs = (7 * smoothedRttMs + sample) / 8
        }
    }
}
